checkstyle-tester/README.md

```diff
@@ -224,6 +224,25 @@ Sevntu's current version must be referenced at https://github.com/checkstyle/con
 
 Finally, just run checkstyle-tester as described above.
 
+## ANTLR Grammar Change Testing Procedure
+
+After modifying Checkstyle's ANTLR grammar, it is neccessary to compare how Checkstyle's checks are affected by your changes.  This is done by comparing the check violation output of your modified branch of Checkstyle against Checkstyle's master branch, and comparing the abstract syntax trees of both branches. 
+
+#### Check Regression Report
+
+To generate the check difference report, you must separate the modules and settings in the [`checkstyle_checks.xml`](https://github.com/checkstyle/checkstyle/blob/master/config/checkstyle_checks.xml) configuration file, found in the checkstyle repository, into separate configuration files (this helps avoid OOM).  The best way to separate the modules is to determine which typically produce the most violations, such as Indentation and WhiteSpace Checks, and put them each into their own configuration file, then separating the rest into files with roughly ten modules each. You may modify all the modules that depend on external files to use default settings.  **For each of the configuration files, you should use the [`my_check.xml`](https://github.com/checkstyle/contribution/blob/master/checkstyle-tester/my_check.xml) file as a base**, and add the modules from `checkstyle_checks.xml` to it. Then `diff.groovy` should be run on all projects in `projects-to-test-on.properties`, using the `diff.groovy` script once for each configuration file.  [See instructions above](https://github.com/checkstyle/contribution/tree/master/checkstyle-tester#basic-difference-report) for "Basic Difference Report".
+
+#### Abstract Syntax Tree Regression Report
+
+ This report is generated using the [`launch_diff_antlr.sh`](https://github.com/checkstyle/contribution/blob/master/checkstyle-tester/launch_diff_antlr.sh) script, and selecting (uncommenting) **only** the checkstyle project from [`projects-to-test-on.properties`](https://github.com/checkstyle/contribution/blob/master/checkstyle-tester/projects-to-test-on.properties).  To use this script, first you must modify the [`launch_diff_variables.sh`](https://github.com/checkstyle/contribution/blob/master/checkstyle-tester/launch_diff_variables.sh) file to reflect the location of the variables that  [`launch_diff_antlr.sh`](https://github.com/checkstyle/contribution/blob/master/checkstyle-tester/launch_diff_antlr.sh) uses to produce the report. Please note that `PULL_REMOTE` will need to be set to the name of the remote repository where your branch resides. If your branch is found at `origin/name-of-your-branch-here`, you will set
```

Comment:
>  **only** the checkstyle project

I would expand this to specifically say the `the checkstyle project sent as the PR`.

This is because master checkstyle won't have inputs for the new tree. So it will truly only be regression, but the PR is required to add new inputs to test the new features, so it will also show regression on the new changes. Some of which will just be exceptions turning into a tree for the full file, but it will atleast allow us to see the new tree and point to issues with it from the reports.

Reply:
I am confused; you're stating that we should amend `projects-to-test-on.properties` to checkout the contributor's PR branch?

Reply:
Your report isn't providing `projects-to-test-on.properties` so I don't know which version of "checkstyle (2)" you used. Even specifying master or a local folder wouldn't give me this information either.

> you're stating that we should amend projects-to-test-on.properties to checkout the contributor's PR branch? 

Yes.

It is not that your report doesn't look correct, I am just saying it may be better to generate the report on the PR. It really comes down to the type of changes you are making and which version of the repo has the most to offer us in the regression report.

If you are adding brand new functionality that can't be parsed as is, Checkstyle master will not have any inputs to cover this case. Running regression on master will show old code is still handled the same way, but it won't show anything new. You will only get half the coverage. You won't see anything new.

If you are fixing existing functionality that is in a corner case, we may not have any inputs to cover this case in master. You won't see any differences in the tree to show the old problem is fixed. You will only see this if you run on the PR version of the repo.

PR checkstyle will always be master plus new input cases that cover more things. When you come down to it, PR may have more benefits to master. The only time this could not be the case is if we remove inputs from master, but that is usually rare.

Reply:
This is an excellent point. I had not considered using the PR branch to test on, this is a great way to incorporate more information into the same report.  Thank you for explaining!

Reply:
No problem, I would add something like this to the read me changes.

One note I didn't hit on, when you add the PR version, you probably need to have it in a completely separate folder separated from the repo. Regression changes branches in the repo you give it. It will check out PR and master, so the files it runs on will change and may cause some issue with the report.

Reply:
Ok, I've updated the README based on this conversation.  In order to simplify what we've discussed, I've just instructed the reader to comment out all projects and append a line to the file that will create a whole new folder and use their PR to run the AST regression test on. Let me know what else I can do here.